apps/src/javalab/Backpack.jsx

```diff
@@ -48,20 +56,74 @@ class Backpack extends Component {
   handleImport = () => {
     const {selectedFiles} = this.state;
     if (selectedFiles.length > 0) {
-      selectedFiles.forEach(filename => {
-        this.props.backpackApi.fetchFile(
-          filename,
-          () => {} /* onError, currently do nothing */,
-          fileContents =>
-            this.props.onImport(filename, fileContents) /* onSuccess */
-        );
-      });
+      this.validateFilenamesForImport(
+        this.importFiles,
+        this.showImportWarning,
+        this.showImportError
+      );
     }
+  };
+
+  importFiles = () => {
+    this.state.selectedFiles.forEach(filename => {
+      this.props.backpackApi.fetchFile(
+        filename,
+        () => {} /* onError, currently do nothing */,
+        fileContents =>
+          this.props.onImport(filename, fileContents) /* onSuccess */
+      );
+    });
     this.collapseDropdown();
   };
 
+  showImportWarning = files => {
+    this.setState({
+      openDialog: Dialog.IMPORT_WARNING,
+      overwriteFileList: files
+    });
+  };
+
+  showImportError = files => {
+    this.setState({
+      openDialog: Dialog.IMPORT_ERROR,
+      overwriteFileList: files
+    });
+  };
+
+  validateFilenamesForImport = (
+    successCallback,
+    warnCallback,
+    errorCallback
+  ) => {
+    let hiddenFilenamesUsed = [];
+    let visibleFilenamesUsed = [];
+    const {selectedFiles} = this.state;
+    const {sources} = this.props;
+    selectedFiles.forEach(filename => {
+      const source = sources[filename];
```

Comment:
Just confirming we don't do any funky string manipulation when files are saved that could result in us not picking up a matching filename (eg, trimming spaces, capitalizing first letter, etc.)?

Reply:
Not that I know of!